v2/health.go

```diff
@@ -0,0 +1,146 @@
+		options.HealthCheckTimeout = options.HealthCheckInterval
+	}
+
+	return &HealthChecker{
+		clients: clients,
+		options: options,
+	}
+}
+
+// StartPeriodicHealthCheck starts the periodic health check for the provided HealthCheckable.
+// The health check will run on each client in the HealthChecker.
+// Stops when the context is canceled.
+func (h *HealthChecker) StartPeriodicHealthCheck(ctx context.Context, hc HealthCheckable) {
+	for namespaceName, client := range h.clients {
+		go func(namespaceName string, client *azservicebus.Client) {
+			h.periodicHealthCheck(ctx, hc, namespaceName, client)
+		}(namespaceName, client)
+	}
+}
+
+func (h *HealthChecker) periodicHealthCheck(ctx context.Context, hc HealthCheckable, namespaceName string, client *azservicebus.Client) {
+	nextCheck := time.Now()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(time.Until(nextCheck)):
+			sbCtx, cancelFunc := context.WithTimeout(ctx, h.options.HealthCheckTimeout)
+			err := hc.HealthCheck(sbCtx, client, cancelFunc)
+			if err != nil {
+				getLogger(ctx).Error(err.Error())
+			}
+			hc.IncHealthCheckMetric(namespaceName, err)
+			nextCheck = nextCheck.Add(h.options.HealthCheckInterval)
+		}
+	}
+}
+
+// HealthCheckable is an interface for performing health checks on azservicebus.Sender and azservicebus.Receiver.
+type HealthCheckable interface {
+	HealthCheck(ctx context.Context, client *azservicebus.Client, cancelFunc context.CancelFunc) error
+	IncHealthCheckMetric(namespaceName string, healthCheckErr error)
+}
+
+// SenderHealthChecker performs health checks on azservicebus.Sender.
+type SenderHealthChecker struct {
+	EntityName string
+}
+
+// HealthCheck performs a health check on azservicebus.Sender by creating a new message batch.
+func (s *SenderHealthChecker) HealthCheck(ctx context.Context, client *azservicebus.Client, cancelFunc context.CancelFunc) error {
+	defer cancelFunc()
+	sbSender, err := client.NewSender(s.EntityName, nil)
+	if err != nil {
+		return err
+	}
+	if _, err = sbSender.NewMessageBatch(ctx, nil); err != nil {
+		return err
+	}
+	return sbSender.Close(ctx)
+}
+
+// IncHealthCheckMetric increments the sender health check metric based on the health check result.
+func (s *SenderHealthChecker) IncHealthCheckMetric(namespaceName string, healthCheckErr error) {
+	if healthCheckErr != nil {
+		sender.Metric.IncHealthCheckFailureCount(namespaceName, s.EntityName)
+	} else {
+		sender.Metric.IncHealthCheckSuccessCount(namespaceName, s.EntityName)
+	}
+}
+
+// ReceiverHealthChecker performs health checks on azservicebus.Receiver.
+type ReceiverHealthChecker struct {
+	EntityName       string
+	SubscriptionName string
+}
+
+// HealthCheck performs a health check on the azservicebus.Receiver by peeking a message.
+func (r *ReceiverHealthChecker) HealthCheck(ctx context.Context, client *azservicebus.Client, cancelFunc context.CancelFunc) error {
+	defer cancelFunc()
+	sbReceiver, err := r.createReceiver(client)
```

Comment:
should we do 

```golang
defer func () { 
  if err := sbReceiver.Close(ctx); err!=nil {
    getLogger(ctx).Errorf(err)
  }
}()
```
just after the create?

wondering about the consequences of the Peek failing, and leaving a receiver not closed.
from the outside, if you get an error, you can't close the receiver.

Reply:
so you need to define a var err outside the defer, and you an still modify its value inside the defer, and have it returned.
see https://stackoverflow.com/a/19934989/156415 for example

so rather than my simplified example:
```golang
var err error
var sbReceiver ...
sbReceiver, err = r.createReceiver(client)
defer func () { 
  err = sbReceiver.Close(ctx); 
}()
return err
```

Reply:
same for the sender case